File: android/app/src/main/java/local/elflix/android/Nachschub.java
package local.elflix.android;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

/**
 * Kommt zu einer abgeschlossenen Serie noch etwas? - der Takt dazu.
 *
 * <h2>Was gefehlt hat</h2>
 *
 * <p>Am Rechner sieht ELFIX seit jeher von Zeit zu Zeit nach, ob zu einer
 * abgeschlossenen Serie neue Folgen erschienen sind: der Titel kommt dann
 * zurück auf die Watchlist, rückt auf die erste ungesehene Folge und bekommt
 * einen Hinweis. Auf Android gab es das nicht - der ganze Vorgang stand in
 * {@code main.js}, und was dort steht, sieht das Telefon nie.
 *
 * <p>Die Folge merkt man erst, wenn kein Rechner mehr läuft. „Black Torch"
 * war am Samstag mit Folge 9 da, und am Fernseher blieb der Watchparty-Titel
 * archiviert, bis irgendwann jemand den PC einschaltete. Wer den Fernseher als
 * einziges Gerät benutzt, bekam den Nachschub überhaupt nie zu sehen.
 *
 * <h2>Was hier steht - und was nicht</h2>
 *
 * <p>Hier steht der <em>Takt</em>: wann ein Durchgang läuft, wie viele Titel
 * er ansieht und wann der nächste dran ist. Die Entscheidung, ob eine Folge
 * Nachschub ist, steht in {@code nachschub.js} im gemeinsamen Kern - derselben
 * Datei, die der Rechner fragt. Es gibt also keine zweite Erkennung, sondern
 * eine, die jetzt auch das Telefon erreicht.
 *
 * <h2>Warum so selten</h2>
 *
 * <p>Jeder geprüfte Titel kostet zwei Seitenaufrufe beim Anbieter. Sechs je
 * Durchgang alle sechs Stunden - dieselben Zahlen wie am Rechner, und sie
 * kommen aus derselben Quelle ({@code nachschub.PRO_LAUF}). Die faire Rotation
 * darüber ({@code nachschub.kandidaten}) sorgt dafür, dass nicht immer
 * dieselben sechs drankommen: sie sortiert nach dem letzten Blick auf den
 * Titel, und der steht am Eintrag.
 *
 * <p>Der Zeitpunkt des letzten Durchgangs liegt daneben in den
 * SharedPreferences. Ein Telefon wird zwanzigmal am Tag gestartet - ohne
 * diesen Stempel liefe bei jedem Start ein Durchgang, und das wäre genau die
 * Dauerabfrage, die niemand will.
 */
public final class Nachschub {
    private static final String TAG = CrashReporter.TAG;

    /** Dieselbe Ablage wie die übrigen Einstellungen der App. */
    static final String ABLAGE = "elflix_settings";
    /** Wann zuletzt ein Durchgang lief - als Millisekunden seit der Epoche. */
    static final String SCHLUESSEL_ZULETZT = "nachschub_zuletzt";

    /**
     * So lange nach dem Start wird gewartet.
     *
     * <p>Ein Start hat Dringenderes zu tun: Anbieter laden, Startseite bauen,
     * Filterlisten prüfen. Der Rechner wartet an derselben Stelle zwanzig
     * Sekunden; hier sind es dreißig, weil ein Fernseher langsamer hochkommt.
     */
    private static final long ANLAUF_MS = 30_000L;

    /** Und in diesem Abstand danach - der Wert aus dem Kern, siehe {@link #taktSetzen}. */
    private static final long RUECKFALL_INTERVALL_MS = 6 * 60 * 60 * 1000L;

    private final Context context;
    private final Bestand bestand;
    private final Handler takt = new Handler(Looper.getMainLooper());

    private long intervallMs = RUECKFALL_INTERVALL_MS;
    private int proLauf = 6;
    private boolean laeuft = false;
    private boolean unterwegs = false;

    private final Runnable durchgang = this::pruefen;

    public Nachschub(Context context, Bestand bestand) {
        this.context = context.getApplicationContext();
        this.bestand = bestand;
    }

    /**
     * Portion und Abstand aus dem Kern übernehmen.
     *
     * <p>Damit stehen die Zahlen an genau einer Stelle. Kommt der Kern nicht
     * dazu, bleibt es bei den Rückfallwerten oben - sie sind dieselben, aber
     * ein Rückfall ist ein Rückfall und keine zweite Quelle.
     */
    public void taktSetzen(int proLauf, long intervallMs) {
        if (proLauf > 0) this.proLauf = proLauf;
        if (intervallMs > 0) this.intervallMs = intervallMs;
    }

    /**
     * Den Takt anwerfen.
     *
     * <p>Der erste Durchgang läuft nach der Anlaufzeit - aber nur, wenn seit
     * dem letzten wirklich genug Zeit vergangen ist. Sonst wird bis zum
     * fälligen Zeitpunkt gewartet.
     */
    public void starten() {
        if (laeuft) return;
        laeuft = true;
        takt.postDelayed(durchgang, Math.max(ANLAUF_MS, verbleibend()));
    }

    /** Beim Beenden: kein Durchgang mehr, und der Handler hält nichts fest. */
    public void anhalten() {
        laeuft = false;
        takt.removeCallbacks(durchgang);
    }

    /** Wie lange es bis zum nächsten fälligen Durchgang noch dauert. */
    private long verbleibend() {
        long zuletzt = ablage().getLong(SCHLUESSEL_ZULETZT, 0L);
        if (zuletzt <= 0) return 0;
        long seither = System.currentTimeMillis() - zuletzt;
        // Eine Uhr, die zurückgestellt wurde, darf den Takt nicht für Wochen
        // anhalten: ein negatives "seither" zählt als "lange her".
        if (seither < 0) return 0;
        return Math.max(0, intervallMs - seither);
    }

    private SharedPreferences ablage() {
        return context.getSharedPreferences(ABLAGE, Context.MODE_PRIVATE);
    }

    private void planen(long inMs) {
        if (!laeuft) return;
        takt.removeCallbacks(durchgang);
        takt.postDelayed(durchgang, Math.max(60_000L, inMs));
    }

    private void pruefen() {
        if (!laeuft || bestand == null) return;
        if (unterwegs) {
            // Der vorige Durchgang wartet noch auf Anbieterseiten. Zwei
            // gleichzeitig wären doppelt so viele Aufrufe für dasselbe
            // Ergebnis.
            planen(intervallMs);
            return;
        }
        long faellig = verbleibend();
        if (faellig > 0) {
            planen(faellig);
            return;
        }
        unterwegs = true;
        // Der Stempel gehört vor den Lauf und nicht dahinter. Bricht der
        // Durchgang ab - kein Netz, App wird weggewischt -, soll der nächste
        // Start nicht sofort wieder einen auslösen.
        ablage().edit().putLong(SCHLUESSEL_ZULETZT, System.currentTimeMillis()).apply();
        bestand.nachschubPruefen(proLauf, gefunden -> {
            unterwegs = false;
            if (gefunden > 0) Log.i(TAG, "Nachschub: " + gefunden + " Titel wieder offen");
            planen(intervallMs);
        });
    }
}

File: StreamingBrowserElectron/src/nachschub.js
"use strict";

/**
 * Nachschub: kommt zu einer abgeschlossenen Serie noch etwas?
 *
 * <h2>Warum das ein eigenes Modul ist</h2>
 *
 * <p>Diese Frage stand vollstaendig in {@code main.js} - Auswahl der Titel,
 * Abruf der Seiten, Vergleich der Grenzen, Reaktivierung. Was in main.js steht,
 * sieht das Telefon nie. Die Folge war eine Schieflage, die man erst merkt,
 * wenn kein Rechner mehr laeuft: "Black Torch" war am Samstag mit Folge 9 da,
 * und auf dem Fernseher blieb der Titel archiviert, bis irgendwann jemand den
 * PC einschaltete.
 *
 * <p>Deshalb liegt die <em>Entscheidung</em> jetzt hier und die
 * <em>Beschaffung</em> dort, wo sie hingehoert: am Rechner holt Electron die
 * Seite, auf dem Telefon holt Java sie (die Kekse der laufenden Sitzung hat nur
 * Java, und die Anbieterseiten liegen hinter Cloudflare). Beide reichen ihren
 * Abruf als {@code holen} herein - dieselbe Bauweise wie {@code kalender.js}
 * und {@code empfehlungslauf.js}.
 *
 * <h2>Was hier <em>nicht</em> noch einmal steht</h2>
 *
 * <p>Nichts von der eigentlichen Rechnung. Die Folgenidentitaet, der Vergleich
 * zweier Staende, die Frage "liegt das hinter dem, was abgeschlossen wurde"
 * und der Weg zur naechsten Folge kommen aus {@code fortschritt.js}
 * ({@link episodeIdentity}, {@link compareEpisodeIdentity},
 * {@link hasNewEpisodeAfterCompletedFavorite}, {@link nextEpisodeAfterFavoriteUrl});
 * die Seiten liest {@code discover.js}. Hier steht nur, in welcher Reihenfolge
 * das gefragt wird und was daraus folgt.
 *
 * <h2>Die Vorsicht, die eingebaut ist</h2>
 *
 * <p>Ein falscher Nachschub ist teurer als ein verpasster. Er reisst einen
 * Titel aus der Mediathek, setzt ihn auf eine Folge, die es nicht gibt, und
 * stellt eine archivierte Watchparty-Runde wieder auf aktiv. Anbieterseiten
 * antworten aber halb, leer, aus einem Zwischenspeicher oder gar nicht.
 *
 * <p>Reaktiviert wird deshalb nur, wenn der neue Umfang <em>strikt hinter</em>
 * zweierlei liegt: hinter der zuletzt bekannten Grenze <em>und</em> hinter der
 * Folge, auf der der Titel abgeschlossen wurde. Alles andere - gleich, kleiner,
 * unbekannt, unlesbar - laesst den Eintrag, wie er ist.
 */

const fortschritt = require("./fortschritt");
const { extractSeriesBounds, extractUnplayableEpisodes } = require("./discover");

/**
 * Wie viele Serien ein Durchgang prueft.
 *
 * <p>Jede kostet zwei Seitenaufrufe. Sechs sind die Zahl, mit der der Rechner
 * seit jeher faehrt; auf dem Telefon gilt dieselbe, damit "in sinnvollen
 * Abstaenden" auf beiden Geraeten dasselbe heisst.
 */
const PRO_LAUF = 6;

/** Und in welchem Abstand. */
const INTERVALL_MS = 6 * 60 * 60 * 1000;

/**
 * Die Serienseite ohne Staffel und Folge - dort stehen alle Staffeln.
 */
function serienSeiteUrl(value) {
  try {
    const url = new URL(String(value || ""));
    url.hash = "";
    url.search = "";
    const pfad = url.pathname
      .replace(/\/(?:episode|folge)-\d+\/?$/i, "")
      .replace(/\/(?:staffel|season)-\d+\/?$/i, "");
    if (pfad === url.pathname) return "";
    url.pathname = pfad;
    return url.href;
  } catch {
    return "";
  }
}

/**
 * Die Staffeluebersicht zu einer Folgenadresse.
 *
 * <p>Auf der Episodenseite steht die Folgenliste nicht - dort ist nicht zu
 * sehen, dass die hinteren Nummern nur Hinweise auf eine zusammengefasste
 * Folge sind. Die Uebersicht weiss es.
 */
function staffelSeiteUrl(value) {
  try {
    const url = new URL(String(value || ""));
    url.hash = "";
    url.search = "";
    const pfad = url.pathname.replace(/\/(?:episode|folge)-\d+\/?$/i, "");
    if (pfad === url.pathname || !/\/(?:staffel|season)-\d+$/i.test(pfad)) return "";
    url.pathname = pfad;
    return url.href;
  } catch {
    return "";
  }
}

/**
 * Welche Titel dieser Durchgang ansieht - und in welcher Reihenfolge.
 *
 * <p><b>Reihum, und nicht immer dieselben.</b> Hier stand einmal die
 * Sortierung nach {@code completedAt}: sechs Titel je Durchgang, und weil die
 * Reihenfolge sich nie aendert, waren es <em>dieselben</em> sechs - bei jedem
 * Durchgang, fuer immer. An einer echten Ablage nachgezaehlt: 18 abgeschlossene
 * Serien mit bekannter Grenze, sechs davon geprueft, zwoelf <b>nie</b>.
 *
 * <p>Sortiert wird deshalb nach dem letzten Blick auf den Titel. Wer noch nie
 * geprueft wurde, steht vorn (kein Stempel = Zeitpunkt null), danach der
 * aelteste Blick; bei Gleichstand entscheidet, was zuletzt zu Ende geschaut
 * wurde.
 *
 * <p>Ausgenommen bleibt ein laufendes Wiederansehen: dort steht der Eintrag auf
 * einer fruehen Folge, und das saehe fuer diese Pruefung aus wie eine Serie,
 * die auf halber Strecke stehengeblieben ist. Nachschub faellt beim naechsten
 * Durchgang auf, wenn der Durchlauf vorbei ist.
 */
function kandidaten(favoriten, hoechstens = PRO_LAUF) {
  const zuletztGeprueft = (favorit) => Date.parse(favorit?.newEpisodeCheckedAt || 0) || 0;
  return (Array.isArray(favoriten) ? favoriten : [])
    .filter((favorit) => favorit && favorit.completed && !favorit.rewatching)
    .filter((favorit) => (favorit.type || fortschritt.inferMediaType(favorit.url)) === "serie")
    .filter((favorit) => fortschritt.sanitizePositiveNumber(favorit.finalSeason)
      && fortschritt.episodeIdentity(favorit.url || ""))
    .sort((links, rechts) => (
      zuletztGeprueft(links) - zuletztGeprueft(rechts)
      || Date.parse(rechts.completedAt || 0) - Date.parse(links.completedAt || 0)
    ))
    .slice(0, Math.max(0, Number(hoechstens) || 0));
}

/**
 * Was die Anbieterseiten ueber den Umfang der Serie sagen.
 *
 * <p>Zwei Seiten, weil eine nicht reicht: die Serienseite nennt die Zahl der
 * Staffeln, die Folgenzahl der <em>letzten</em> steht erst auf deren eigener
 * Seite. Die Staffeluebersicht wird dabei doppelt gelesen - einmal auf die
 * hoechste Folgennummer ({@code extractSeriesBounds}), einmal auf die letzte
 * wirklich <em>abspielbare</em> ({@code extractUnplayableEpisodes}). Die
 * zweite geht vor: eine Nummer, die nur als "in E18 enthalten" dasteht, ist
 * kein Nachschub.
 *
 * @param seiten {{serienHtml: string, staffelHtml: string, staffel: number}}
 * @returns {{seasons: number, episodes: number}|null} - null, wenn sich nicht
 *          einmal die Staffelzahl lesen laesst. Eine leere, halbe oder kaputte
 *          Antwort landet genau hier und fuehrt zu keiner Aenderung.
 */
function umfangLesen(seiten) {
  const serienHtml = String(seiten?.serienHtml || "");
  if (!serienHtml) return null;
  const staffeln = fortschritt.sanitizePositiveNumber(extractSeriesBounds(serienHtml).seasons);
  if (!staffeln) return null;

  const staffelHtml = String(seiten?.staffelHtml || "");
  const inStaffel = staffelHtml ? extractSeriesBounds(staffelHtml, staffeln) : null;
  const gelistet = staffelHtml ? extractUnplayableEpisodes(staffelHtml) : null;
  const folgen = fortschritt.sanitizePositiveNumber(gelistet?.lastPlayable)
    || fortschritt.sanitizePositiveNumber(gelistet?.listed)
    || fortschritt.sanitizePositiveNumber(inStaffel?.episodes);
  return { seasons: staffeln, episodes: folgen };
}

/**
 * Die Entscheidung. Genau diese Funktion fragen Rechner und Telefon.
 *
 * <p>Sie fasst nichts an und redet mit niemandem: herein kommt der Eintrag und
 * was die Seiten hergaben, heraus kommt, was daraus folgt. Deshalb laesst sie
 * sich auf beiden Plattformen mit denselben Eingaben pruefen - und deshalb
 * kann es keine zwei Meinungen darueber geben, ob eine Folge Nachschub ist.
 *
 * <p><b>Die beiden Riegel.</b> Der neue Umfang muss strikt hinter der zuletzt
 * bekannten Grenze liegen (sonst waere jede halb geladene Seite ein Anlass) und
 * strikt hinter der Folge, auf der der Titel abgeschlossen wurde (sonst holte
 * eine gewachsene Staffelzahl einen Titel nach vorn, dessen Ende laengst
 * dahinter liegt). Die zweite Frage beantwortet
 * {@code hasNewEpisodeAfterCompletedFavorite} - dieselbe Funktion, mit der die
 * Seitengrenzen beim Schauen entscheiden.
 *
 * @param favorit der abgeschlossene Eintrag
 * @param umfang  {{seasons, episodes}} aus {@link umfangLesen}
 * @returns {{art: "nichts"|"neu", grund: string, aenderung?: object, label?: string}}
 */
function nachschubUrteil(favorit, umfang) {
  const nichts = (grund) => ({ art: "nichts", grund });
  if (!favorit || !umfang) return nichts("kein-umfang");
  if (!favorit.completed || favorit.rewatching) return nichts("nicht-abgeschlossen");

  const identity = fortschritt.episodeIdentity(favorit.url || "");
  if (!identity) return nichts("keine-folge");

  const bekannteStaffel = fortschritt.sanitizePositiveNumber(favorit.finalSeason);
  const bekannteFolge = fortschritt.sanitizePositiveNumber(favorit.finalEpisode);
  if (!bekannteStaffel) return nichts("grenze-unbekannt");

  const neueStaffeln = fortschritt.sanitizePositiveNumber(umfang.seasons);
  const neueFolgen = fortschritt.sanitizePositiveNumber(umfang.episodes);
  if (!neueStaffeln) return nichts("umfang-unlesbar");

  // Eine neue Staffel braucht keine Folgenzahl; eine neue Folge derselben
  // Staffel braucht beide Zahlen, sonst wuerde aus "unbekannt" ein Fortschritt.
  const neueStaffel = neueStaffeln > bekannteStaffel;
  const neueFolge = neueStaffeln === bekannteStaffel
    && bekannteFolge > 0
    && neueFolgen > bekannteFolge;
  if (!neueStaffel && !neueFolge) {
    return nichts(neueStaffeln < bekannteStaffel || (neueStaffeln === bekannteStaffel && neueFolgen < bekannteFolge)
      ? "umfang-geschrumpft"
      : "nichts-neues");
  }

  // Zweiter Riegel: liegt der neue Stand auch hinter der Folge, die hier
  // abgehakt wurde? Bei einer neuen Staffel ist die Folgenzahl der neuen
  // Staffel noch unbekannt oder klein (S2E1 < S1E12 waere sonst ein
  // Rueckschritt) - verglichen wird deshalb ueber die Staffel, und die ist
  // groesser. Die geteilte Regel macht genau diesen Unterschied.
  const schluessel = identity.key || "";
  const vorher = { key: schluessel, season: bekannteStaffel, episode: bekannteFolge };
  const jetzt = {
    key: schluessel,
    season: neueStaffeln,
    episode: neueStaffel ? Math.max(1, neueFolgen) : neueFolgen
  };
  if (!fortschritt.hasNewEpisodeAfterCompletedFavorite(favorit, vorher, jetzt)) {
    return nichts("nicht-hinter-dem-abschluss");
  }

  // Und wohin. Erste Wahl ist die geteilte Regel; sie zaehlt vom Stand des
  // Eintrags aus und nicht von der gespeicherten Grenze - bei einem Eintrag,
  // der auf einer frueheren Folge steht als seine Grenze behauptet, ist das
  // die richtigere Antwort.
  const ziel = fortschritt.nextEpisodeAfterFavoriteUrl(favorit, neueStaffeln, jetzt.episode)
    || (neueStaffel
      ? fortschritt.replaceEpisodeUrl(favorit.url, bekannteStaffel + 1, 1)
      : fortschritt.replaceEpisodeUrl(favorit.url, bekannteStaffel, bekannteFolge + 1));
  if (!ziel || ziel === favorit.url) return nichts("kein-ziel");

  const zielIdentity = fortschritt.episodeIdentity(ziel);
  const imRaum = Boolean(favorit.watchpartyRoom);
  const aenderung = {
    url: ziel,
    normalizedUrl: fortschritt.normalizeFavoriteUrl(ziel),
    season: zielIdentity?.season || favorit.season || 0,
    episode: zielIdentity?.episode || favorit.episode || 0,
    finalSeason: neueStaffeln,
    finalEpisode: neueFolgen || fortschritt.sanitizePositiveNumber(favorit.finalEpisode),
    completed: false,
    completedAt: "",
    completedManually: false,
    rewatching: false,
    episodeCompleted: false,
    continuePending: true,
    hideFromContinueWatching: false,
    progress: 0,
    position: 0,
    currentTime: 0,
    duration: 0
  };

  // Auf die Merkliste kommt nur, was einem selbst gehoert. Ein Raum-Eintrag
  // gehoert seiner Runde - die private Watchlist entsteht nie aus einer
  // Watchparty, und umgekehrt holt genau dieser Fund den archivierten
  // Raumtitel zurueck in "Gemeinsam weiterschauen".
  if (imRaum) aenderung.watchpartyArchived = false;
  else aenderung.favorite = true;

  const label = neueStaffel
    ? `Staffel ${aenderung.season} ist da`
    : `Folge ${aenderung.episode} ist da`;
  return { art: "neu", grund: neueStaffel ? "neue-staffel" : "neue-folge", aenderung, label };
}

/**
 * Ein Durchgang - die Verkabelung um die Entscheidung herum.
 *
 * <p>Was hier passiert, ist auf beiden Geraeten dasselbe und steht deshalb
 * auch nur einmal da: Kandidaten waehlen, je Kandidat zwei Seiten holen,
 * {@link nachschubUrteil} fragen, das Ergebnis auf den Eintrag legen.
 *
 * <p><b>Gestempelt wird der Versuch und nicht der Erfolg.</b> Ein Titel, dessen
 * Seite gerade nicht antwortet, stuende sonst beim naechsten Durchgang wieder
 * ganz vorn und verstopfte die Runde fuer alle anderen - genau die Verstopfung,
 * die die Sortierung in {@link kandidaten} beheben soll.
 *
 * @param holen     {(url) => Promise<{html: string}|null>} der Abruf der
 *                  Plattform. Ein Fehler darf hier geworfen werden; er wird
 *                  gefangen und zaehlt wie "keine Antwort".
 * @param protokoll optional, bekommt eine Zeile je Fund
 * @param jetzt     optional, fuer die Pruefungen
 */
function erstellen({ holen, protokoll = null, jetzt = () => new Date() } = {}) {
  async function seite(url) {
    if (!url) return "";
    try {
      const antwort = await holen(url);
      return String(antwort?.html || "");
    } catch {
      return "";
    }
  }

  async function umfangLaden(favorit) {
    const serienHtml = await seite(serienSeiteUrl(favorit.url));
    if (!serienHtml) return null;
    const staffeln = fortschritt.sanitizePositiveNumber(extractSeriesBounds(serienHtml).seasons);
    if (!staffeln) return null;
    // Die Folgenzahl der letzten Staffel steht auf deren eigener Seite.
    const letzteUrl = fortschritt.replaceEpisodeUrl(favorit.url, staffeln, 1);
    const staffelHtml = letzteUrl
      ? await seite(staffelSeiteUrl(letzteUrl) || letzteUrl)
      : "";
    return umfangLesen({ serienHtml, staffelHtml, staffel: staffeln });
  }

  /**
   * @returns {{geprueft: Array, gefunden: Array, geaendert: boolean}} -
   *          `gefunden` sind die Eintraege, die in <em>diesem</em> Durchgang
   *          neu dazugekommen sind. Nur ueber die wird gemeldet: `newEpisodeAt`
   *          bleibt am Eintrag stehen, bis der Titel geoeffnet wird, und von
   *          dort gemeldet kaeme dieselbe Meldung immer wieder.
   */
  async function lauf(favoriten, hoechstens = PRO_LAUF) {
    const liste = kandidaten(favoriten, hoechstens);
    const geprueft = [];
    const gefunden = [];
    let geaendert = false;

    for (const favorit of liste) {
      favorit.newEpisodeCheckedAt = jetzt().toISOString();
      geaendert = true;
      geprueft.push(favorit);

      const umfang = await umfangLaden(favorit);
      const urteil = nachschubUrteil(favorit, umfang);
      if (urteil.art !== "neu") continue;

      Object.assign(favorit, urteil.aenderung);
      favorit.newEpisodeAt = jetzt().toISOString();
      favorit.newEpisodeLabel = urteil.label;
      gefunden.push(favorit);
      if (protokoll) protokoll(`${favorit.title || favorit.url}: ${urteil.label}`);
    }
    return { geprueft, gefunden, geaendert };
  }

  return { lauf, umfangLaden };
}

module.exports = {
  erstellen,
  kandidaten,
  nachschubUrteil,
  umfangLesen,
  serienSeiteUrl,
  staffelSeiteUrl,
  PRO_LAUF,
  INTERVALL_MS
};
